src/main/java/frc/robot/subsystems/Elevator.java

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import edu.wpi.first.math.controller.ProfiledPIDController;
+import edu.wpi.first.math.trajectory.TrapezoidProfile;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj2.command.CommandScheduler;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants.ElevatorConstants;
+
+public class Elevator extends SubsystemBase {
+  private final CANSparkMax elevatorMotor;
+  private final DigitalInput limitSwitch;
+  private final TrapezoidProfile.Constraints constraints =
+      new TrapezoidProfile.Constraints(1.75, 0.75);
+  private final ProfiledPIDController pidController =
+      new ProfiledPIDController(ElevatorConstants.kP, 0, 0, constraints, 0);
+  private final Encoder encoder;
+
+  public Elevator(int elevatorMotorID, int encoderChannelA, int encoderChannelB, int limitSwitchChannel) {
+    this.elevatorMotor = new CANSparkMax(elevatorMotorID, MotorType.kBrushless);
+    elevatorMotor.setInverted(true);
+    elevatorMotor.setIdleMode(IdleMode.kBrake);
+    this.encoder = new Encoder(encoderChannelA, encoderChannelB);
+    encoder.setDistancePerPulse(1.0 / 360.0 * 2.0 * Math.PI * 1.5);
+    this.limitSwitch = new DigitalInput(limitSwitchChannel);
+    CommandScheduler.getInstance().registerSubsystem(this);
+  }
+
+  public void setGoal(double position) {
+    pidController.setGoal(position);
+  }
+
+  public boolean getLimitSwitchValue() {
+    return this.limitSwitch.get();
+  }
+
+  public void extend(double speed) {
+    elevatorMotor.set(speed);
+  }
+
+  public double getExtension() {
+    return encoder.getDistance();
+  }
+
+  public void periodic() {
+    elevatorMotor.set(pidController.calculate(encoder.getDistance()));
+  }
```

Comment:
Keep in mind that this won't actually control the elevator if the current command sets the motor speeds directly.  The only time this method will do anything is if the current command uses `setGoal` or doesn't control the elevator at all (in which case the elevator will suddenly attempt to go to whatever the most recent setpoint was, which could be dangerous)

Reply:
The idea is to use this to keep the elevator from falling due to gravity -- would it work if I used setGoal in the end of each elevator command?

Reply:
Yes... but you'd have to remember to put it in every command that controls the elevator.  And since the elevator has a default command that sets the motor speed (typically just to zero, since the joysticks won't be used often), the periodic method will only end up doing something when a command is scheduled and sets the elevator's *setpoint*, not motor speeds, doesn't set the setpoint in every run iteration, and doesn't finish quickly (basically, sets the setpoint on initialize, does nothing in `run`, and only finishes when the subsystem hits the setpoint).  That would be a terrible way to write a command, so basically the `periodic` method in the subsystem will do nothing.

Note that the docs mention that the periodic method and default commands have similar functions - so having both a `periodic` method *and* a default command is redundant.

Reply:
Periodic methods are generally used for pushing monitoring data to smartdashboard or other similar non-motion tasks. A command is more appropriate for anything that controls robot motion, even if that motion is "stop"